### ISSUE TITLE
perf: is device_put / TPU-launch overhead actually a scheduler bottleneck? (host-vs-forward masking analysis)

### ISSUE THREAD

**Issue by Yajingleo** [sgl-project/sglang-jax#1500]
## Context

Following up on #293 (is the scheduler thread the throughput bottleneck?) and the discussion on #1484. On current `main`, the scheduler's per-step host work is **masked** under the TPU forward. In #1484 @JamesBrianD noted the current bottlenecks are instead:
1. `device_put` when converting `ScheduleBatch` → `ForwardBatch`, and
2. TPU JIT launch overhead.

This issue tracks a focused question: **in which regime do `device_put` / launch overhead actually bite** — i.e. cause measurable TPU idle / lost throughput — since where they're *masked* they cost ~0.

## What `device_put` is

`ForwardBatch.init_new` (`forward_batch_info.py`) H2D-transfers ~8 per-step batch-metadata arrays (`input_ids`, `seq_lens`, `cache_loc`, `positions`, …) via `device_array` → `jax.make_array_from_callback`, sharded on the `data` (DP) axis. It runs every forward step and allocates fresh device buffers each step (no reuse). Optimization lever (if it un-masks): preallocate persistent bucketed-shape device buffers + in-place update.

## Findings — nnodes=1 (single host)

Setup: Qwen3-8B, v6e-4, conc 16, 1024/1024, JAX profiler; **device busy/step = forward-module compute ÷ unprofiled step** (profiler-invariant).

| regime | forward | throughput | device idle |
|--------|---------|-----------|-------------|
| DP1/TP4 | 4.85 ms | 3296 tok/s | ~2% |
| DP2/TP2 | 7.35 ms | 2176 tok/s | ~2–6% |
| DP4/TP1 | 14.5 ms | 1104 tok/s | ~3–6% |

**The device is ~saturated in every DP regime → `device_put`/launch is masked under the forward → reducing it buys ~0 throughput here.** The 2–6% residual is within measurement noise (chip-to-chip spread 3% vs 6% for identical work).

Note the DP-scaling intuition (more DP → more `device_put` → un-mask) is *disproved* on a single host: with fixed chips, more DP means less TP-per-replica, which **lengthens the forward faster than it grows the host**, so the device only gets *more* saturated.

A local micro-benchmark of the one *new* multi-host host cost — the per-iteration pickle + ZMQ broadcast of the request list from node 0 — is ~0.24 µs in steady state (no new arrivals) and ~1 ms even on a 64-request admission burst, i.e. also plausibly masked.

## Multi-host (nnodes>1) — in progress

Architecturally different (worth calling out): multi-controller, node 0 is the leader that pickle+ZMQ-broadcasts the request list every loop; every node **replays the full scheduling** (so `build`/`device_put` is *replicated* per host, not split); the forward is a **synchronous collective** (slowest host's step gates it). So the multi-host-specific risk is likely **sync-collective jitter → tail latency**, not mean throughput. I'm currently measuring `nnodes=2` on a v6e-16 (2 hosts × 8 chips) and will post the device-idle result here.

## Question

@JamesBrianD — to make sure we chase the right regime: what config did your profiling show the `device_put`/launch overhead in?
- `nnodes` / slice size (single-host vs multi-host)?
- concurrency and model (esp. large MoE)?
- was it a **throughput** regression or a **tail-latency** one?

Happy to share the profiling scripts (JAX-profiler device-idle analyzer) if useful.

**Comment by Yajingleo** [comment]
## Multi-host result (nnodes=4) — the device is NOT saturated here (~13% idle)

Ran the profiler on a **v6e-16 = 4 hosts × 4 chips**, `--nnodes 4 --tp-size 16 --dp-size 4` (one 4-chip DP replica per host, so each host runs the same TP4 workload as the single-host DP1/TP4 baseline), conc 64.

```
baseline throughput = 11,126 tok/s (conc 64)  ->  unprofiled step = 5.75 ms
forward (device compute) median      = 5.02 ms   (rank-0's 4 chips, all ~identical)
device busy/step = 5.02 / 5.75 = 87%  ->  device idle ≈ 13%
```

**This is the first regime with meaningful device idle.** Recap of the sweep:

| regime | forward | step | device idle |
|--------|---------|------|-------------|
| single-host DP1/TP4 | 4.85 ms | 4.85 ms | ~2% |
| single-host DP2/TP2 | 7.35 ms | 7.35 ms | ~2–6% |
| single-host DP4/TP1 | 14.5 ms | 14.5 ms | ~3–6% |
| **multi-host nnodes=4, tp16/dp4** | **5.02 ms** | **5.75 ms** | **~13%** |

The `device idle` is `1 − forward_compute / unprofiled_step`, both profiler-invariant, so it's a steady-state number, not a tracing artifact.

### Interpretation
Each host runs an *independent* DP replica (TP4 on its local chips — no cross-host tensor all-reduce), yet the per-step wall time is ~0.9 ms longer than the identical single-host workload. That ~13% gap is **multi-host coordination overhead**: the forward is one synchronized SPMD collective across all 4 hosts, so each step carries a cross-host barrier, and the per-host request pickle-broadcast + replicated `get_model_worker_batch`/`device_put` now sit on the critical path of that synchronized launch. Unlike single-host (device saturated), **the device is ~13% idle → there is real recoverable headroom in multi-host**, consistent with `device_put`/launch overhead mattering *here* rather than single-host.

### Caveats
- This is **nnodes=4**, not 2 — on this platform `v6e-8` is single-host, so v6e-16 (4 hosts) is the smallest multi-host slice.
- The **size** of the gap (~13%) is solid; **attributing** it precisely (sync barrier vs broadcast vs replicated build vs launch dispatch) needs a finer host-lane breakdown, and some fraction is likely inherent multi-host sync latency (not all recoverable).

### Question still stands
@JamesBrianD — was your `device_put`/launch observation on a multi-host slice? If so this lines up, and the next step is a host-lane trace to split the ~13% into recoverable (build/broadcast/launch) vs inherent (sync) — happy to do that.
